Enable flake8 and re-enable it everywhere

QuLogic opened the pull request:
## PR Summary

Due to the removal of the examples symlink in #11141, pytest was no longer seeing the examples and running pep8 on them. This adds that directory back to pytest's search.

It also adds a rebase of #10940 to enable flake8.

## PR Checklist

- [N/A] Has Pytest style unit tests
- [x] Code is PEP 8 compliant
- [N/A] New features are documented, with examples if plot related
- [N/A] Documentation is sphinx and numpydoc compliant
- [N/A] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [N/A] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

QuLogic commented:
The meaning of the additional exceptions to flake8 options are described [here](https://github.com/matplotlib/matplotlib/pull/10940#issuecomment-387396163) by @cclauss.

anntzer commented on pytest.ini:
```diff
@@ -7,8 +7,9 @@ markers =
 
-pep8ignore =
-    * E111 E114 E115 E116 E121 E122 E123 E124 E125 E126 E127 E128 E129 E131 E226 E240 E241 E242 E243 E244 E245 E246 E247 E248 E249 E265 E266 E704 W503
+flake8-max-line-length = 80
```
Shouldn't this be left at 79 (the default, i.e. remove this line)?

QuLogic replied:
I got this from @cclauss's PR, but I think you're right, so I'm going to remove it.

tacaswell replied:
sorry, off-by-one bug on my part.

cclauss commented:
LOL... https://github.com/matplotlib/matplotlib/pull/10940#issuecomment-377939400

QuLogic commented:
Well, okay, now pyside(2) seems to be broken.

Also, I have no idea why, but flake8 make the tests ridiculously slow. I ran a non-parallel test locally and it took ~7 minutes, but with the `--flake8` flag it's not finished for the past hour.

timhoffm commented:
Maybe related https://github.com/tholo/pytest-flake8/issues/19?

anntzer commented:
If https://github.com/matplotlib/matplotlib/pull/11477#issuecomment-399709807 is still an issue, perhaps replace pytest-flake8 by a direct invocation of flake8? (with e.g. https://github.com/snoack/flake8-per-file-ignores for per-file ignores)
(TBH I don't really see the point of a pytest plugin that just calls another executable (if it is not active by default).)

jklymak commented:
I ran 
- `pytest -v --pep8 lib/matplotlib/tests/test_axes.py`: 143.09 s
- `pytest -v --flake8 lib/matplotlib/tests/test_axes.py`: 131.09 s

Sure I didn't have the exceptions turned on or anything (the pep8 test failed).

QuLogic commented:
Well, I tried conda, virtualenv, and even downloading the docker image that Travis used, but I cannot reproduce the slow checks anymore. I guess standalone `flake8` is the only option; the only annoyance is that configuration will have to go into a new file, as `pytest.ini` is not a supported file.

cclauss commented:
From http://flake8.pycqa.org/en/latest/user/invocation.html

flake8  --config=CONFIG       Path to the config file that will be the authoritative
                        config source. This will cause Flake8 to ignore all
                        other configuration files.

tacaswell commented:
Is the issue that pytest spins up a new flake8 process for each file which just takes forever on the travis vms?

tacaswell commented:
watching a running test it is not hung, just from 91% on it seems to be taking it's time (and then we hit the 50 minute timeout).

cclauss commented:
One minute, 10 seconds seems reasonable to me.

$ __time flake8 . --config=pytest.ini__  # Python 3.7 on Travis CI
```
real	0m26.769s
user	1m10.204s
sys	0m0.953s
```

QuLogic commented:
I decided to just use `.flake8` instead, for a few reasons:
* I needed to rewrite all the lines to use commas anyway;
* `flake8-per-file-ignores` now produces a warning for unnecessary exceptions, and quite a few had to be removed;
* this works with my editor linter out-of-the-box.

I added some new ignores for things that occur several times:
* N801: 30
* N802: 366
* N803: 283
* N806: 973
* N812: 6

and fixed several other things that occur only a couple of times.

QuLogic commented:
At some point in the near future, I would like to get rid of the `**` exceptions, as they cause confusion when linting a single file (`per-file-ignores` thinks that the exception is unnecessary and raises its `X100` error).

jklymak commented:
Where does the flake8 output show up in the test?  I can't see any evidence that it ran, but maybe I'm looking in the wrong spot...

QuLogic commented:
That's because it passed :wink: I'll add a message to say it's okay on there.

cclauss commented:
Do the flake8 tests first and the pytests after?

jklymak commented:
@cclauss  then it won't fail if flake8 fails?

#11597 shows that @tacaswell is indeed correct - that failing a test still returns a positive Travis check the way this is setup now....

jklymak commented:
#11597 shows that adding `set -e` causes a failure when pytest should fail.  Testing if it fails if there is a flake8 error

QuLogic commented on .travis.yml:
```diff
@@ -160,6 +160,9 @@ before_script: |
 script: |
   echo "Calling pytest with the following arguments: $PYTEST_ADDOPTS"
   python -mpytest
+  if [[ $RUN_FLAKE8 == 1 ]]; then
```
I don't like `set -e`; it bleeds into Travis' code and causes spurious errors, so you need to remember to turn it off. The last time that happened, we ended up just removing it instead. I split these into separate entries instead, which does fail correctly.

QuLogic commented:
Splitting the two commands into separate entries [should work fine](https://travis-ci.org/QuLogic/matplotlib/builds/401489734).

QuLogic commented on examples/units/basic_units.py:
```diff
@@ -25,13 +25,13 @@ def __get__(self, obj, objtype=None):
 
 
 class TaggedValueMeta(type):
-    def __init__(cls, name, bases, dict):
```
It may be that the object is a class/type here, but it's still an instance at this point, so I'm not sure calling it something other than `self` is warranted.

tacaswell commented:
Thanks everyone!
